### DISCUSSION TITLE
Passing argument to an inlined container ? (in quest of modularity)

### DISCUSSION THREAD

**Discussion by EricBoix** [argoproj-labs/hera#727] [Q&A] [answered]
I'm trying to pass arguments to a container build within an explicit Step.
My Hera (version 5.1.3) code, that tries to blend/mix
* the [Steps Types](https://hera.readthedocs.io/en/stable/examples/workflows/steps_types/) Hera example
* with the [Steps Inline Workflow](https://hera.readthedocs.io/en/stable/examples/workflows/upstream/steps_inline_workflow/) "upstream" example 

goes

```python
from hera.workflows import Container, Parameter, Step, Steps, Workflow

whalesay_container = Container(
    name="whalesay",
    inputs=[Parameter(name="message")],
    image="docker/whalesay",
    command=["cowsay"],
    args=["{{inputs.parameters.message}}"],
)

my_step = Step(                                 # Explicit Step construction
    name="inlined-step",
    inline=whalesay_container,            # Has to use "inline" because "template" was failing
    arguments=[Parameter(name="message", value="Inline.")],
)

cluster = define_cluster()
with Workflow(generate_name="workflow-name-", entrypoint="main") as w:
    with Steps(name="main") as s:
        whalesay_container(                 # OK: implicit Step construction
            name="not-inlined-step",
            arguments=[Parameter(name="message", value="OK.")],
        ) 
        s.sub_steps.append(my_step).        # FAIL

w.create()
```
The above will fail with a message of the form
```bash
'templates.main.steps failed to resolve {{inputs.parameters.message}}'}
```

Hence my question: how can one pass arguments to an explicitly build Step ?

## Notes
* I fell on the above difficulty when trying to achieve some form modularity when writing my hera workflows. More precisely my need (that follows the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)) is to encapsulate a sequence of container calls (say 4-5 containers triggering a database, as daemon, and checking that DB initialization is indeed correct and that the DB is properly usable) into some convenient/concise callable expression. The above is a trial to achieve modularity through a combination `Container()`, with `Step()` and thus the `inline` form of the `Step` constructor.
* The modularity need is better satisfied with `model.TemplateRef()` (refer to e.g. [this example](https://github.com/argoproj-labs/hera/discussions/732)).

**Comment by elliotgunton** [comment, marked as answer]
This is a problem with Argo itself with `inline` steps being broken before v3.4.9. See https://github.com/argoproj/argo-workflows/pull/11074. This code should work on 3.4.9

**Reply by EricBoix** [reply]
Thanks @elliotgunton . I'm indeed still running on Argo version 3.4.4.
